proposals/configurable_tooling.md

```diff
@@ -0,0 +1,111 @@
+
+## Goals
+- Provide a data scientist's CLI that:
+  1. Allows the data scientist to easily create a project or switch between projects.
+  1. Is able to discover available services such as dashboards, jobs, job histories, notebooks.
+  1. Is easily installed on the client machine.
+  1. Provides common kubeflow operations as customizable subcommands (train, serve, notebook)
+  1. Allows a data scientist or contributor to add or override subcommands using python
+  1. Allows subcommands to be added to only those granted particular privileges
+  1. Enables a subcommand registry   
+
+## Non-Goals
+The CLI subcommands suggested do not proscribe against defining additional subcommands in other proposals.
+
+## UI or API
+1. A CLI with an initial base set of subcommands:    
+   1. General    
+      1. `kf create [-h] [-d] [-r repo] [-b branch] [-i] <namespace>`
+      1. `kf delete [-h] <namespace>`
+      1. `kf remove [-h] <component>`
+      1. `kf run [-h] <component> [container] -- [command]`
+      1. `kf status [-h] [list|health|info][<component>]`
+      1. `kf use [-h] <namespace>`
+      1. `kf whoami [-h]`
+   1. Specific to Kubeflow components but not addressed per se within this proposal    
+      1. `kf notebook [-h] [ls|start|stop] <name>`
+      1. `kf serve [-h] <name> <yaml>`
+      1. `kf shell [-h] <component> [<container>]`
+      1. `kf tensorboard [-h] <name> <yaml>`
+      1. `kf train [-h] <name> <yaml>`
+1. A CLI that is extensible via a subcommand called `command`    
+  1. `kf command [-h] [add|remove|replace|upgrade|list|describe] <subcommand>`
+1. Subcommands call a Kubeless Function (written in python|bash|golang) which:    
+   1. Runs in a container that includes requisite tooling and dependencies like ks, kubeclt, kubeless, python2.7|python3.6
+
+![CLI calls a serverless Function](diagrams/CLI2Function.png)
+## Design
+A data scientist's CLI is a flexible set of commands that are executed against a cluster of DL components. Within kubeflow, this set of components will vary and will use `ks` environments that are parameterized. A CLI with a fixed set of subcommands can not accommodate a configurable set of runtime components provided by kubeflow, of which many may be provided by open source contributors. The kubeflow flexible component design mandates a matching CLI design where subcommands can be provided by component authors or other contributors that provide one or more subcommands for a component. For example different contributors may provide a pytorch CLI that can be used with the pytorch component. The author of the pytorch component may provide their own CLI. Different CLI bindings are possible for any component. Each CLI subcommand invocation will delegate its execution to a serverless Function. This design enables a CLI ecosystem where different functionalities not possible within a fixed CLI deliverable are possible.  
+
+### Implementation
```

Comment:
Do we want FaaS? We are as strong as our weakest dependency, and FaaS is still very early in the game (imho). I'd be more inclined to use native k8s as much as possible and try to limit external projects.
Besides stability of external projects, each one also adds complexity to deployment and operations. Look how much issues adding StorageClass requirement caused...